sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2785,6 +2785,41 @@ class DataFrameSuite extends QueryTest
     val df1 = df.select("a").orderBy("b").orderBy("all")
     checkAnswer(df1, Seq(Row(1), Row(4)))
   }
+
+  test("SPARK-53401: repartitionById - should partition rows to the specified " +
+    "partition ID") {
+    val numPartitions = 10
+    val df = spark.range(100).withColumn("p_id", col("id") % numPartitions)
+
+    val repartitioned = df.repartitionById(numPartitions, $"p_id")
+    val result = repartitioned.withColumn("actual_p_id", spark_partition_id())
+
+    assert(result.filter(col("p_id") =!= col("actual_p_id")).count() == 0)
+
+    assert(result.rdd.getNumPartitions == numPartitions)
+  }
+
+  test("SPARK-53401: repartitionById - should fail when partition ID is null") {
+    val df = spark.range(10).withColumn("p_id",
+      when(col("id") < 5, col("id")).otherwise(lit(null).cast("long"))
+    )
+    val repartitioned = df.repartitionById(5, $"p_id")
+
+    val e = intercept[SparkException] {
+      repartitioned.collect()
+    }
+    assert(e.getCause.isInstanceOf[IllegalArgumentException])
+    assert(e.getCause.getMessage.contains("The partition ID expression must not be null."))
+  }
+
+  test("SPARK-53401: repartitionById - should fail analysis for non-integral types") {
+    val df = spark.range(5).withColumn("s", lit("a"))
+    val e = intercept[AnalysisException] {
+      df.repartitionById(5, $"s").collect()
+    }
+    // Should fail with type error from DirectShufflePartitionID expression
+    assert(e.getMessage.contains("requires an integral type"))
```

Comment:
where do we throw this error now?

Reply:
oh it's execution time. We should fail earlier in the analysis time, like `CheckAnalysis`